README
                                clasp-2.x
              A conflict-driven nogood learning answer set solver
                     http://www.cs.uni-potsdam.de/clasp/
					  http://potassco.sourceforge.net/

OVERVIEW
  clasp is an answer set solver for (extended) normal logic programs.
  It combines the high-level modeling capacities of answer set programming
  with state-of-the-art techniques from the area of Boolean constraint solving.
  The primary clasp algorithm relies on conflict-driven nogood learning,
  a technique that proved very successful for satisfiability checking (SAT).
  Starting with version 2.0, clasp supports parallel (multithreaded) solving.

  clasp is written in (mostly) Standard-C++. It was successfully built and run
  under Linux (x86-32, x86-64) using gcc and Windows (x86-32, x86-64) using
  either Microsoft Visual Studio or MinGW.

  Detailed information (including a User's manual), source code,
  and pre-compiled binaries are available at: http://potassco.sourceforge.net/

LICENSE
  clasp is part of the Potassco project hosted at SourceForge.
  It is distributed under the GNU Public Licens. See COPYING for
  details regarding the license.

PACKAGE CONTENTS
  COPYING      - GNU Public License
  CHANGES      - Major changes between versions
  README       - This file
  configure.{sh,bat}
               - Simple script that creates Makefiles for building clasp (library and application)
  app/         - Source code directory of the command-line interface
  libclasp/    - Directory of the clasp (static) library (sources, documentation, unit tests)
  libprogram_opts/
               - Library for parsing command-line options (needed by app)
  build_vc/    - Directory containing Visual Studio project files for building clasp
  tools/       - Some additional files

BUILDING & INSTALLING
  The preferred way to build clasp is to use make and the provided configure script.
  You'll need to have the GNU Compiler Collection (GCC) version 3 or
  better installed in order to build clasp. You'll also need GNU make 3.80 or better.
  On Microsoft Windows, we recommend using MinGW available from http://www.mingw.org/
  You may want to visit http://www.mingw.org/wiki/Getting_Started for detailed
  instructions on installing MinGW. Make sure to also install MinGW-make.

  In the following it is assumed that 'make' is an alias for the installed GNU make.
  If this is not the case on your system, replace 'make' with the name of the GNU make
  executable (e.g. gmake). Furthermore, on Microsoft Windows use ./configure.bat instead of
  ./configure.sh.

  clasp's multithread support requires the Intel Threading Building Blocks library (version >= 3.x)
  which is freely available at: http://threadingbuildingblocks.org/
  After downloading and installing you may want to set and export the
  TBB30_INSTALL_DIR environment variable.

  Type
    ./configure.sh --help
  to get an overview of all supported build configurations/options.

  To build clasp:
    ./configure.sh
    cd build/release
    make

  To build clasp with multithread support using TBB30_INSTALL_DIR:
    ./configure.sh --with-mt
    cd build/release_mt
    make

  To build clasp with multithread support using custom directory structure:
    ./configure.sh --with-mt TBB_INCLUDE=<path_to_tbb_include> TBB_LIB=<path_to_tbb_lib>
    cd build/release_mt
    make

  To install clasp:
    make install

  By default, 'make install' will install clasp in '/usr/local/bin'
  You can specify an installation prefix other than '/usr/local'
  by running the configure-script with the option '--prefix=PATH'.
  Alternatively, use option '--bindir=PATH' to directly specify the
  installation path.

  Finally, you can always skip installation and simply copy the
  clasp executable to a directory of your choice.

BUILDING WITH Microsoft Visual Studio
  In the directory build_vc/ we provide Microsoft Visual Studio project files
  for building clasp. You can download the freely available express edition
  of Visual C++ from here:
  http://www.microsoft.com/express/Downloads/
  Once installed:
    - open build_vc\vc9\clasp\clasp.sln
    - select the desired solution configuration (typically release_static)
    - build the "app" project

USAGE
  clasp reads problem instances either from stdin, e.g
    cat problem | clasp
  or from a given file, e.g
    clasp problem

  Beside logic programs in SMODELS-format, clasp also supports SAT-problems in DIMACS-,
  Max-SAT in (extended) DIMACS-, and PB-problems in OPB and WBO-format.

  Type
    clasp --help
  to get an overview of the various options supported by clasp.

  In addition to printing status information, clasp also
  provides information about the computation via its exit status.
  The exit status is:
    10: if the problem was found to be satisfiable
    20: if the problem was proved to be unsatisfiable
    0 : if the satisfiability of problem is not known,
        because search was either interrupted or not started
    127: if clasp ran out of memory
    Furthermore, the exit status of 1 indicates an error.


CHANGES
clasp 2.1.2:
  * fixed regression: Inconsistent behaviour in dimacs front-end when running
                      with or without option "--number=1"
clasp 2.1.1: Thursday, 22th November 2012
  * fixed regression: MaxSAT-frontend must not assert pure literals / relaxation variables
  * fixed: spurious warning when using option "--no-lookback"
  * fixed: "--shuffle" and "--counter-restarts" not working as intended with hierarchical optimization
  * fixed: Backslashes in JSON output not correctly quoted

clasp 2.1.0: Monday, 27th August 2012
  * restructured help and revised options and their defaults
    * added "--help[={1..3}]" for showing basic, advanced, or all options
	* added "--configuration" for selecting between prefabricated default configurations
	* revised deletion options and grouped them via common prefix "--del-"
	* revised thread options and added options for size-/topology-based lemma exchange
	* added support for dynamic restarts via "--restarts=D,<args>"
	* added "--sign-def" for setting default sign used in decision heuristic
	* added "--sign-fix" for disabling sign-heuristic
	* added "--init-moms" for enabling/disabling initialization of heuristic with MOMS-score
    * added "--fast-exit" for forcing app exit without cleanup
	* added "--update-act" for enabling LBD-based activity bumping
	* added "--update-mode" for configuring when update/integrate messages are handled in parallel-search
	* added "--learn-explicit" to disable usage of implication graph for learnt constraints
	* added "--share" for configuring physical constraint sharing
	* added "--init-watches" for controlling initialization of watched literals
	* added "--counter-restarts" and "--counter-bump" for enabling counter implication restarts
	* added "--lemma-out-lbd" for restricting lemmas written via "--lemma-out"
	* added "--lemma-in-lbd" for setting initial lbd of lemmas read via "--lemma-in"
	* replaced "--loops-in-heu" with more general "--score-other"
	* replaced "--rand-watches" with more general "--init-watches"
	* replaced "--initial-lookahead" with "--lookahead=<type>[,<num>]"
	* replaced "--recursive-str" with "--strengthen=<mode>[,<antes>]"
	* replaced "--copy-problem" with more general "--share"
	* removed options "--brave", "--cautious", "--solution-recording": all superseded by "--enum-mode=<type>"
	* option "--restarts" now always requires a type;
      and the "limit" parameter now sets the (initial) length of the sequence
  * disbaled signal handling during printing of models
  * fixed: Overflow on parsing large opt-values (issue 3528852)
  * fixed: Parallel bt-enumerator enumerated duplicate models under certain conditions
  * fixed: Sat-Preprocessor failed to expand models correctly under very rare conditions
  * fixed: Erroneous interplay between "otfs=2" and "reverse-arcs" could lead to the
           unjustified removal of problem constraints
  * fixed: Incorrect ordering-predicate in MinimizeConstraint could lead to wrong optima
  * fixed: Solve loop failed to handle restart on total assignment correctly resulting
           in an infinite loop on some optimization problems
  * fixed: Duplicate key in JSON-Output of lemma stats

clasp 2.0.6: Tuesday, 3rd April 2012
  * added better error detection to portfolio parser and fixed a bug
    in the initialization of thread configs
  * first attempt at decoupling learnt db growing from restart schedule using
    independent grow schedule configurable via (hidden) option "--dgrowS"
  * added "--update-lbd=2" to enable usage of "strict" lbds
  * added (experimental) support for counter implication restarts and dynamic restarts
    via hidden options
  * fixed bugs in code for shared clause integration that could led to unpleasant
    results ranging from duplicate/missing models in enumeration to crashing if
    option "--otfs" was used

clasp 2.0.5: Sunday, 29th January 2012
  * fixed: Input parser for opb and (w)cnf failed to read 64bit integers.
  * fixed: Outer bound of inner-outer-restarts sometimes remained constant because of integer arithmetic.
  * fixed: Minimize constraint must not update optimum in enumeration-mode (opt-all).
           Bug introduced in version 2.0.3
  * fixed: Minimize constraint must not update initial (user-set) bound until a model is found.
           Bug introduced in version 2.0.3
  * fixed: Right hand side of minimize constraint not correctly initialized if "--opt-hierarch" was used.
  * fixed: SharedMinimizeData assumed monotonicity of individual levels - could fail
           after merging complementary literals in multi-level minimize constraints.
           Added extra "adjustment"-values to represent values resulting from such merges.

clasp 2.0.4: Tuesday, 29th November 2011
  * some cleanup in help output
    * moved general ASP-specific options to separate group
    * moved "--opt-sat" to basic options
    * simplified default command-line and made "input-dependent" defaults more explicit
  * updated interface spec of ProgramBuilder to make it more robust w.r.t incremental update
  * fixed parsing of "--global-restarts"
  * fixed: default value for option "--rand-watches" was not applied
  * fixed: Enumerator did not broadcast last model/optimum of current GP in parallel search
  * fixed: clauses simplified to binary/ternary are no longer counted twice in stats
  * fixed: eq-preprocessor sometimes did not reorder head list after merging two bodies
  * fixed: dimacs front-end failed to read empty dimacs file

clasp 2.0.3: Tuesday, 23rd August 2011
  * added option "--enum-mode" for setting enumeration algorithm and
    deprecated superseded options "--solution-recording", "--brave", and "cautious"
  * made "--enum-mode=record" the default for optimization
  * added default portfolio settable via "--portfolio=default"
  * added support for arithmetic restarts
  * moved "--opt-heuristic" to "Search Options" to allow for different settings in portfolios

configure.sh
#!/bin/bash

# OPTIONS
claspre=0
static=0
machine=0
mt=0
rpath=0
PREFIX="/usr/local"
BIN_DIR=""
CONFIG=""
# CONFIGURATION
CXXFLAGS="${CXXFLAGS}"
LDFLAGS="${LDFLAGS}"
LDLIBS=""
BUILDPATH=""
INSTALLPATH=""
CXX=""
POST_BUILD=""
TBB_INCLUDE=""
TBB_LIB=""
while [[ $# > 0 ]]; do
	case $1 in
		CXXFLAGS=*) CXXFLAGS=`echo "$1"| sed  's/^[A-Z]*=*//'`
			;;
		LDFLAGS=*) LDFLAGS=`echo "$1"| sed  's/^[A-Z]*=*//'`
			;;
		CXX=*) CXX=`echo "$1"| sed  's/^[A-Z]*=*//'`
			;;
		TBB_INCLUDE=*) TBB_INCLUDE=`echo "$1"| sed  's/^[A-Z_]*=*//'`
			;;
		TBB_LIB=*) TBB_LIB=`echo "$1"| sed  's/^[A-Z_]*=*//'`
			;;
		"--strip")
			POST_BUILD="strip"
			;;
		"--static")
			static=1
			;;
		"--with-claspre")
			claspre=1
			;;
		"--with-mt")
			mt=1
			;;
		"--set-rpath")
			rpath=1
			;;
		"--m32")
			machine=32
			;;
		"--m64")
			machine=64
			;;
		--bindir*|--prefix*|--config*)
			T=`echo "$1"| sed 's/^--[a-z-]*=*//'`
			A=$1
			if [ -z "$T" ]; then
				if [ -z "$2" ]; then
					echo "error: required parameter missing after '$1'"
					exit 1
				fi
				T=$2
				shift
			fi
			case $A in
				--bindir*) BIN_DIR=$T;;
				--prefix*) PREFIX=$T;;
				--config*) CONFIG=$T;;
			esac
			;;
		"--clean")
			rm -rf build/
			exit 0
			;;
		"--help")
			echo
			echo "$0 [options]"
			echo
			echo "  --help             : show this help"
			echo "  --prefix=PREFIX    : set install prefix to PREFIX"
			echo "    Default: '/usr/local'"
			echo "  --bindir=PATH      : set install path to PATH"
			echo "    Default: '${PREFIX}/bin'"
			echo
			echo "  --config=NAME      : set configuration to NAME"
			echo "    NAME=release     : configure for optimized release version"
			echo "    NAME=debug       : configure for debug version"
			echo "    NAME=check       : configure for release version with assertions enabled"
			echo "    <NAME>           : configure for custom configuration with name <NAME>"
			echo
			echo "  --with-mt          : enable multi-thread support (see below)"
			echo "  --set-rpath        : store path to shared libraries in binary header"
			echo "  --with-claspre     : enable claspre features in clasp"
			echo "  --static           : link statically (if supported)"
			echo "  --m32              : force 32-bit binary (if supported)"
			echo "  --m64              : force 64-bit binary (if supported)"
			echo "  --strip            : discard symbols (calls strip after build)"
			echo "  --clean            : remove all generated files"
			echo
			echo "Note: Multi-thread support currently requires Intel® Threading Building Blocks >= 3.x."
			echo "      Use option --with-mt and either set TBB30_INSTALL_DIR environment variable or"
			echo "      explicitly set include and/or library path via:"
			echo "  $0 --with-mt TBB_INCLUDE=<path_to_tbb_include> TBB_LIB=<path_to_tbb_lib>"
			echo
			echo "Note: To create a custom configuration call $0 like this: "
			echo "  $0 --config=my_config CXX=my_gcc CXXFLAGS=my_cxxflags LDFLAGS=my_ldflags"
			echo
			exit 0
			;;
		*)
			echo "*** Error: unknown option $1"
			echo "type '$0 --help' for an overview of supported options"
			exit 1
	esac
	shift
done
if [ -z "$CONFIG" ]; then
	CONFIG="release"
fi
case $CONFIG in
	release) CXXFLAGS="-O3 -DNDEBUG" ;;
	debug)   CXXFLAGS="-g -D_DEBUG -DDEBUG -O1" ;;
	check)   CXXFLAGS="-O2 -DDEBUG" ;;
	*)
		if [ -z "$CXXFLAGS" ]; then
			CXXFLAGS="-O3 -DNDEBUG"
		fi
		;;
esac

BUILDPATH="build/${CONFIG}"

if [[ $mt == 0 ]]; then
	CXXFLAGS="${CXXFLAGS} -DWITH_THREADS=0"
else
	# try to find tbb headers
	echo -ne "Checking for TBB include path..."
	for i in "$TBB_INCLUDE" "$TBB30_INSTALL_DIR/include" "/opt/intel/tbb/include" "/usr/local/include" "/usr/include" "/opt/local/include"; do
		TBB_INCLUDE=""
		if [ -f "$i/tbb/tbb.h" ]; then
			TBB_INCLUDE="$i"
			echo "$TBB_INCLUDE"
			break
		fi
	done
	if [ -z "$TBB_INCLUDE" ]; then
		echo "FAIL"
		echo "*** Error: TBB include path not set!"
		echo "use '$0 TBB_INCLUDE=<path_to_tbb_include>'"
		exit 1
	fi
	# try to find tbb lib
	echo -ne "Checking for TBB library path..."
	for i in "$TBB_LIB" "$TBB30_INSTALL_DIR/lib" "/opt/intel/tbb/lib" "/usr/local/lib" "/usr/lib" "/opt/local/lib" ; do
		TBB_LIB=""
		if [ -f "$i/libtbb.so" -o -f "$i/libtbb.dylib" ]; then
			TBB_LIB="$i"
			echo "$TBB_LIB"
			break
		fi
	done
	if [ -z "$TBB_LIB" ]; then
		echo "FAIL"
		echo "*** Error: TBB library path not set or 'libtbb.{so,dylib}' not found!"
		echo "use '$0 TBB_LIB=<path_to_tbb_library>'"
		exit 1
	fi
	CXXFLAGS="${CXXFLAGS} -DWITH_THREADS=1 -I\"${TBB_INCLUDE}\""
	LDFLAGS="${LDFLAGS} -L\"${TBB_LIB}\""
	LDLIBS="-ltbb"
	if [[ $rpath == 1 ]]; then
		LDFLAGS="${LDFLAGS} -Xlinker \"--rpath=\"${TBB_LIB}\"\""
	fi
	BUILDPATH="${BUILDPATH}_mt"
fi

CXXFLAGS="${CXXFLAGS} -DWITH_CLASPRE=${claspre}"
if [[ $static == 1 ]]; then
	LDFLAGS="${LDFLAGS} -static"
	BUILDPATH="${BUILDPATH}_static"
fi
if [[ $machine != 0 ]]; then
	LDFLAGS="${LDFLAGS} -m${machine}"
	CXXFLAGS="${CXXFLAGS} -m${machine}"
	BUILDPATH="${BUILDPATH}_m${machine}"
fi
if [ -z "$BIN_DIR" ]; then
	INSTALLPATH="${PREFIX}/bin"
else
	INSTALLPATH=$BIN_DIR
fi

# create & prepare build hierarchy
ROOTPATH="../.."
LIB_CLASP="libclasp"
LIB_OPTS="libprogram_opts"
mkdir -p "$BUILDPATH/app"
mkdir -p "$BUILDPATH/bin"
mkdir -p "$BUILDPATH/$LIB_CLASP/lib"
mkdir -p "$BUILDPATH/$LIB_OPTS/lib"
cd "$BUILDPATH"
rm -f  .CONFIG $LIB_CLASP/.CONFIG $LIB_OPTS/.CONFIG
rm -f  Makefile $LIB_CLASP/Makefile $LIB_OPTS/Makefile
rm -f FLAGS
# write FLAGS
touch FLAGS
if [ ! -z "$CXX" ]; then
echo "CXX         := ${CXX}"      >> FLAGS
else
echo "CXX         ?= g++"         >> FLAGS
fi
echo "CXXFLAGS    := ${CXXFLAGS}" >> FLAGS
echo "WARNFLAGS   := -W -Wall"    >> FLAGS
echo "LDFLAGS     := ${LDFLAGS}"  >> FLAGS
echo ""                           >> FLAGS

# create Makefiles
LIB_MAKES="${ROOTPATH}/tools/Base.in ${ROOTPATH}/tools/LibRule.in ${ROOTPATH}/tools/BaseRule.in"
PRO_MAKES="${ROOTPATH}/tools/Base.in ${ROOTPATH}/tools/ProjRule.in ${ROOTPATH}/tools/BaseRule.in"
cat $LIB_MAKES >> $LIB_CLASP/Makefile
cat $LIB_MAKES >> $LIB_OPTS/Makefile
cat $PRO_MAKES >> Makefile
# write project config
touch  .CONFIG
echo "PROJECT_ROOT := $ROOTPATH"            >> .CONFIG
echo "TARGET       := bin/clasp"            >> .CONFIG
echo "FLAGS        := FLAGS"                >> .CONFIG
echo "SOURCE_DIR   := \$(PROJECT_ROOT)/app" >> .CONFIG
echo "INCLUDE_DIR  := \$(PROJECT_ROOT)/app" >> .CONFIG
echo "OUT_DIR      := app"                  >> .CONFIG
echo "INCLUDES     := -I\$(PROJECT_ROOT)/${LIB_CLASP} -I\$(PROJECT_ROOT)/${LIB_OPTS}" >> .CONFIG
echo "SUBDIRS      := ${LIB_CLASP} ${LIB_OPTS}"  >> .CONFIG
echo "LIBS         := ${LIB_CLASP}/lib/${LIB_CLASP}.a ${LIB_OPTS}/lib/${LIB_OPTS}.a" >> .CONFIG
echo "LDLIBS       := ${LDLIBS}"            >> .CONFIG
echo "INSTALL_DIR  := \"${INSTALLPATH}\""   >> .CONFIG
if [ ! -z "$POST_BUILD" ]; then
echo "POST_BUILD  := $POST_BUILD"           >> .CONFIG
fi
echo ""                                     >> .CONFIG
# write lib configs
touch $LIB_CLASP/.CONFIG $LIB_OPTS/.CONFIG
echo "PROJECT_ROOT := ${ROOTPATH}/.."      >> $LIB_CLASP/.CONFIG
echo "PROJECT_ROOT := ${ROOTPATH}/.."      >> $LIB_OPTS/.CONFIG
echo "TARGET       := lib/${LIB_CLASP}.a"  >> $LIB_CLASP/.CONFIG
echo "TARGET       := lib/${LIB_OPTS}.a"   >> $LIB_OPTS/.CONFIG
echo "FLAGS        := ../FLAGS"            >> $LIB_CLASP/.CONFIG
echo "FLAGS        := ../FLAGS"            >> $LIB_OPTS/.CONFIG
echo "SOURCE_DIR   := \$(PROJECT_ROOT)/${LIB_CLASP}/src" >> $LIB_CLASP/.CONFIG
echo "TEST_DIR     := \$(PROJECT_ROOT)/${LIB_CLASP}/tests" >> $LIB_CLASP/.CONFIG
echo "TEST_TARGET  := ./test-lib"          >> $LIB_CLASP/.CONFIG
echo "SOURCE_DIR   := \$(PROJECT_ROOT)/${LIB_OPTS}/src" >> $LIB_OPTS/.CONFIG
echo "INCLUDE_DIR  := \$(PROJECT_ROOT)/${LIB_CLASP}/clasp" >> $LIB_CLASP/.CONFIG
echo "INCLUDE_DIR  := \$(PROJECT_ROOT)/${LIB_OPTS}/program_opts" >> $LIB_OPTS/.CONFIG
echo "INCLUDES     := -I\$(PROJECT_ROOT)/${LIB_CLASP}" >> $LIB_CLASP/.CONFIG
echo "INCLUDES     := -I\$(PROJECT_ROOT)/${LIB_OPTS}" >> $LIB_OPTS/.CONFIG
echo "" >> $LIB_CLASP/.CONFIG
echo "" >> $LIB_OPTS/.CONFIG

# DONE
echo
echo "Configuration successfully written to ${BUILDPATH}."
echo "Make flags written to ${BUILDPATH}/FLAGS."
echo
echo "To compile clasp type:"
echo "  cd ${BUILDPATH}"
echo "  make"
echo
echo "To install clasp afterwards type:"
echo "  make install"
echo "or copy '${BUILDPATH}/clasp' to a directory of your choice."
if [ ! -d "$INSTALLPATH" ]; then
echo
echo "Note: install path '$INSTALLPATH' does not exist"
echo "      'make install' will fail unless it is first created!"
fi
echo
echo "Note: \"make\" must correspond to GNU Make 3.8 or later."
echo
